### PR TITLE
Adding Test inject in scenario and simulations

### PR DESCRIPTION
**Pull request by johanah29**
#1106 MVP 1,2 and 3

### CODE REVIEW

File: openbas-front/src/admin/components/common/injects/InjectPopover.tsx [hunk 1]
```diff
@@ -6,21 +6,23 @@ import Transition from '../../../../components/common/Transition';
...
 import type { Inject, InjectStatus, InjectStatusExecution, Tag } from '../../../../utils/api-types';
-import { duplicateInjectForExercise, duplicateInjectForScenario, tryInject } from '../../../../actions/Inject';
+import { duplicateInjectForExercise, duplicateInjectForScenario, tryInject, testInject } from '../../../../actions/Inject';
 import { useAppDispatch } from '../../../../utils/hooks';
 import DialogDuplicate from '../../../../components/common/DialogDuplicate';
 
 interface Props {
-  inject: InjectStore;
+  inject: InjectStore & { inject_test?: boolean }; // FIXME: Inject object coming from multiple endpoints with difeerent properties
   tagsMap: Record<string, Tag>;
```
Comment (savacano28) [inline openbas-front/src/admin/components/common/injects/InjectPopover.tsx:15]: difeerent

File: openbas-api/src/main/java/io/openbas/rest/inject/output/InjectOutput.java [hunk 2]
```diff
@@ -59,6 +61,11 @@ public class InjectOutput {
   @JsonProperty("inject_content")
   private ObjectNode content;
 
+  @JsonProperty("inject_test")
```
Comment (RomuDeuxfois) [inline openbas-api/src/main/java/io/openbas/rest/inject/output/InjectOutput.java:64]: inject_testable is maybe more relevant
What do you think ?

File: openbas-api/src/main/java/io/openbas/rest/inject/InjectApi.java [hunk 3]
```diff
@@ -64,511 +66,554 @@
...
+  }
+
+  @GetMapping(INJECT_URI + "/test/{injectId}")
+  public InjectStatus testInject(@PathVariable String injectId) {
+    Inject inject = injectRepository.findById(injectId).orElseThrow();
+    User user = this.userRepository.findById(currentUser().getId()).orElseThrow();
+    List<ExecutionContext> userInjectContexts = List.of(
+        this.executionContextService.executionContext(user, inject, "Direct test")
+    );
+    io.openbas.execution.Injector executor = context.getBean(
```
Comment (RomuDeuxfois) [inline openbas-api/src/main/java/io/openbas/rest/inject/InjectApi.java:176]: io.openbas.execution.Injector should be replaced by the imported package
Injector

File: openbas-front/src/actions/Inject.js [hunk 4]
```diff
@@ -13,6 +13,11 @@ export const tryInject = (injectId) => (dispatch) => {
   return getReferential(null, uri, null)(dispatch);
 };
 
+export const testInject = (injectId) => (dispatch) => {
```
Comment (RomuDeuxfois) [inline openbas-front/src/actions/Inject.js:16]: Why are you using dispatch here ?

File: openbas-front/src/admin/components/common/injects/InjectPopover.tsx [hunk 5]
```diff
@@ -6,21 +6,23 @@ import Transition from '../../../../components/common/Transition';
...
   setSelectedInjectId: (injectId: Inject['inject_id']) => void;
   isDisabled: boolean;
+  canBeTested: boolean;
 }
 
 const InjectPopover: FunctionComponent<Props> = ({
   inject,
   setSelectedInjectId,
   isDisabled,
+  canBeTested,
```
Comment (RomuDeuxfois) [inline openbas-front/src/admin/components/common/injects/InjectPopover.tsx:25]: You can add a default value here
canBeTested = false

File: openbas-front/src/admin/components/simulations/simulation/timeline/TimelineOverview.tsx [hunk 6]
```diff
@@ -201,11 +201,12 @@ const TimelineOverview = () => {
                                 {fndt(inject.inject_date)}
                               </div>
                             </div>
-                                }
+                          }
                         />
                         <ListItemSecondaryAction>
                           <InjectPopover
                             inject={inject}
+                            canBeTested={false}
```
Comment (RomuDeuxfois) [inline openbas-front/src/admin/components/simulations/simulation/timeline/TimelineOverview.tsx:209]: You remove this if you have a default value in the InjectPopover component

File: openbas-model/src/main/java/io/openbas/database/model/InjectStatus.java [hunk 7]
```diff
@@ -86,16 +86,42 @@ public static InjectStatus fromExecution(Execution execution, Inject executedInj
...
     ExecutionStatus finalStatus = numberOfError > 0 && numberOfSuccess > 0 ? ExecutionStatus.PARTIAL : globalStatus;
     injectStatus.setName(execution.isAsync() ? ExecutionStatus.PENDING : finalStatus);
     injectStatus.setTrackingEndDate(Instant.now());
-    injectStatus.setTrackingTotalExecutionTime(Duration.between(injectStatus.getTrackingSentDate(), injectStatus.getTrackingEndDate()).getSeconds());
+    injectStatus.setTrackingTotalExecutionTime(
+        Duration.between(injectStatus.getTrackingSentDate(), injectStatus.getTrackingEndDate()).getSeconds());
+    return injectStatus;
+  }
+
+  public static InjectStatus fromExecutionTest(Execution execution) {
```
Comment (RomuDeuxfois) [inline openbas-model/src/main/java/io/openbas/database/model/InjectStatus.java:106]: There are some duplicate code here
You can mutualize a bunch of lines

**Comment by RomuDeuxfois** [comment]
I can test an inject when there are no content
It should not be possible

image

**Comment by johanah29** [comment]
image
"Test" option is now disabled when no team is set.

File: openbas-front/src/actions/injects/Inject.d.ts [hunk 8]
```diff
@@ -9,9 +9,8 @@ export type InjectInput = {
   inject_depends_duration_seconds: number;
 };
 
-export type InjectStore = Omit<Inject, 'inject_tags' | 'inject_content' | 'inject_injector_contract' | 'inject_teams'> & {
+export type InjectStore = Omit<Inject, 'inject_tags' | 'inject_content' | 'inject_injector_contract'> & {
```
Comment (RomuDeuxfois) [inline openbas-front/src/actions/injects/Inject.d.ts:12]: Why are you removing this ?

**Comment by RomuDeuxfois** [comment]
One last thing:
one we click on test, the popover should close
image

File: openbas-front/src/actions/injects/Inject.d.ts (hunk 8 above)
Comment (johanah29) [reply openbas-front/src/actions/injects/Inject.d.ts:12]: I need to access inject_teams to enable/disable the Test action

**Comment by johanah29** [comment]
image

**Review by RomuDeuxfois — APPROVED** [review]

**Pull request merged**